personalities/sctfe/storage/bbolt/dedup.go

```diff
@@ -0,0 +1,179 @@
+}
+
+// NewStorage returns a new BBolt storage instance with a dedup and size bucket.
+//
+// The dedup bucket stores <leafID, idx> pairs.
+// The size bucket has a single entry: <"size", X>, where X is the largest contiguous index from 0
+// that has been inserted in the deudp bucket.
+//
+// If a database already exists at the provided path, NewStorage will load it.
+func NewStorage(path string) (*Storage, error) {
+	// TODO(better logging message)
+	db, err := bolt.Open(path, 0600, nil)
+	if err != nil {
+		return nil, fmt.Errorf("bolt.Open(): %v", err)
+	}
+	s := &Storage{db: db}
+
+	err = db.Update(func(tx *bolt.Tx) error {
+		dedupB := tx.Bucket([]byte(dedupBucket))
+		sizeB := tx.Bucket([]byte(sizeBucket))
+		if dedupB == nil && sizeB == nil {
+			klog.V(2).Infof("NewStorage: no pre-existing buckets, will create %q and %q.", dedupBucket, sizeBucket)
+			_, err := tx.CreateBucket([]byte(dedupBucket))
+			if err != nil {
+				return fmt.Errorf("create %q bucket: %v", dedupBucket, err)
+			}
+			sb, err := tx.CreateBucket([]byte(sizeBucket))
+			if err != nil {
+				return fmt.Errorf("create %q bucket: %v", sizeBucket, err)
+			}
+			klog.V(2).Infof("NewStorage: initializing %q with size 0.", sizeBucket)
+			err = sb.Put([]byte("size"), itob(0))
+			if err != nil {
+				return fmt.Errorf("error reading logsize: %v", err)
+			}
+		} else if dedupB == nil && sizeB != nil {
+			return fmt.Errorf("inconsistent deduplication storage state %q is nil but %q it not nil", dedupBucket, sizeBucket)
+		} else if dedupB != nil && sizeB == nil {
+			return fmt.Errorf("inconsistent deduplication storage state, %q is not nil but %q is nil", dedupBucket, sizeBucket)
+		} else {
+			klog.V(2).Infof("NewStorage: found pre-existing %q and %q buckets.", dedupBucket, sizeBucket)
+		}
+		return nil
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("error initializing buckets: %v", err)
+	}
+
+	return s, nil
+}
+
+// Add inserts entries in the dedup bucket and updates the size bucket if need be.
+//
+// If an entry is already stored under a given key, Add does not update it, even if the new value
+// is different.
+// The context is here for consistency with interfaces, but isn't used by BBolt.
+func (s *Storage) Add(_ context.Context, kvs []dedup.KV) error {
+	for _, kv := range kvs {
+		err := s.db.Update(func(tx *bolt.Tx) error {
+			db := tx.Bucket([]byte(dedupBucket))
+			sb := tx.Bucket([]byte(sizeBucket))
+			sizeB := sb.Get([]byte("size"))
+			if sizeB == nil {
+				return fmt.Errorf("can't find log size in bucket %q", sizeBucket)
+			}
+			size := btoi(sizeB)
+
+			if old := db.Get(kv.K); old != nil {
+				klog.V(3).Infof("Add(): bucket %q already contains an entry for %q, not updating", dedupBucket, hex.EncodeToString(kv.K))
+			}
+			if err := db.Put(kv.K, itob(kv.V)); err != nil {
+				return err
+			}
+			// sizeB is indexes from 1 since it's a size, kv.K from 0.
+			// Therefore, if they're equal, kv is a new entry.
+			if size == kv.V {
```

Comment:
Hmm, how does this work if we end up calling something like:

- Add(1) (put: Size=1)
- Add(2) (put: Size=2)
- Add(4) (no put, 4 != 2+1)
- Add(3) (put: Size=3)
- Add(5) (no put, 5 != 3+1)
- Add(6) (no put, 6 != 3+1)
- ...

I _suspect_ we'll get stuck with Size=3 forever, but maybe I've missed something?

What you've got is probably fine for the "track the log" part of updating, but currently I _think_ the CTFE handler will end up calling in here with the assigned index before the tracker gets to see it.

Reply:
Ah yes well found. For added context, I added this no-overriding feature last minute, and didn't pay attention.

Other than fixing this small bug, there are three ways this could go:
  - update the index if the one from the log is smaller. We'd converge faster on long term consistency through-ought restarts, at the cost of short term consistency.
  - never call AddCertIndex, only rely on the background synchronisation feature. That would prevent issues like this, but would also make deduplication less effective (you ned to do a round of sync before a server can dedup an entry it has just processed, and is probably the most likely server to process this exact entry again). Also, that would make it quite different from a GlobalDeduplication implementation. Thoughts?

Reply:
Actually, In the example that you give, Add has is being called out of order, therefore the size should not be updated, this is working as intended.
Add can be called by Sync, or when a leaf is added locally. Let's call "Add called by Sync"  "AddFromSync":
 - AddFromSync(1) (put: Size=1)
 - AddFromSync(2) (put: Size=2)
 - Add(4) (no put, 4 != 2+1) <-- This call cannot from sync, otherwise AddFromSync(3) would have been called first.
 - AddFromSync(3) (put: Size=3)
 - Add(5) (no put, 5 != 3+1)
 - Add(6) (no put, 6 != 3+1)
 - AddFromSync(4) (put: Size=4)